drafts/homemade-signals.md

```diff
@@ -0,0 +1,232 @@
+
+Effects are the powerhouses that make everything move, as they produce a side effect every time a source Signal changes.
+That allows frameworks to re-render relevant segments of the UI much more efficiently than elaborate change detectors or by comparing virtual DOM trees.
+
+And now, off to building Signals ourselves.
+
+## Basic Reactive State
+
+The first element we're implementing is the elementary building block: a Signal.
+This initial version is just a simple wrapper around a value with accessor functions.
+
+```js
+function signal(initialValue) {
+    let _value = initialValue;
+    return {
+        get value() {
+            return _value;
+        },
+        set value(newValue) {
+            if (newValue !== _value) {
+                _value = newValue;
+            }
+        }
+    }
+}
+```
+
+It doesn't notify anyone reading the value yet.
+But we can pass its reference around and always get the up-to-date value whenever we read it.
+That is so far only useful for primitive values since JavaScript otherwise passes them by value.
+
+```js
+const counter = signal(0);
+counter.value += 1;
+console.log(counter.value); // 1
+```
+
+## Derived Values
+
+Signals that are derived, or _computed_, from other Signals are an importation use case.
+We want to be able to use them like this:
+
+```js
+const counter = signal(0);
+const double = computed(()= => counter.value * 2);
+```
+
+Due to their calculated nature, we only need them to be readable, not writable.
+We want it to re-run the given computation every time we read it.
+A very simple version could look something like the following:
+
+```js
+function computed(computation) {
+    return {
+        get value() {
+            return computation();
+        }
+    }
+}
+```
+
+So far we get the current value if we ask for it (pull principle), but nothing happens after a Signal changes.
+
+## Fuel the Engine with Effects
+
+Side effects are essentially the part that makes Signals move and brings our applications to life.
+Unlike computed Signals, an effect doesn't return anything.
+We're not assigning effects to some variable but want it to _do_ something for us every time there's a change in the dependency tree.
+The most basic variation could look like this:
+
+```js
+function effect(fn) {
+    fn();
+}
+effect(() => console.log(`${counter.value} * 2 = ${double.value}`));
+```
+
+This would run the function once and print `0 * 2 = 0`.
+But how are we now making everything reactive?
+We need some way to subscribe to changes in the source Signals without adding anything to the API.
+This eliminates the (often in examples used) option of a `subscribe` method, as it defeats the purpose of ergonomic, minimalistic reactive values.
+So we need to look at it from the other side.
+Whenever a Signal's value is requested, it has to know who asks for it.
+JavaScript doesn't have any built-in mechanism to inspect the call stack, which means we track it ourselves.
+
+```js
+let caller;
+```
```

Comment:
does this bring any value to the article? It just confused me ^^

Reply:
aha this is a global variable. This wasnt clear to me, since after the code block.
You could write something like: We are going to add a global variable `let caller;`. This global variabel....